File: uvaonlinejudge/10004 - Bicoloring.c
#include <stdio.h>
#define MAX_NODES 199

int read_test_case(int adj_list[MAX_NODES][MAX_NODES + 1], int *n){
    int i, l, u, v;
    scanf("%d", n);
    if(!*n)
        return 0;
    for(i = 0; i < *n; i++){
        adj_list[i][0] = 0;
    }
    scanf("%d", &l);
    for(i = 0; i < l; i++){
        scanf("%d %d", &u, &v);
        adj_list[u][++adj_list[u][0]] = v;
        adj_list[v][++adj_list[v][0]] = u;        
    }
    return 1;
}

int is_bicolored(int adj_list[MAX_NODES][MAX_NODES + 1], int n){
    int i, u, v, colors[MAX_NODES] = {1};
    int stack[MAX_NODES], stack_size = 1;
    stack[0] = 0;
    while(stack_size > 0){
        v = stack[--stack_size];
        for(i = 1; i <= adj_list[v][0]; i++){
            u = adj_list[v][i];
            if(colors[u] == 0){
                colors[u] = 3 - colors[v];
                stack[stack_size++] = u;
            }else if(colors[u] == colors[v]){
                return 0;
            }
        }
    }
    return 1;
}

int main(){
    int n, adj_list[MAX_NODES][MAX_NODES + 1];
    while(read_test_case(adj_list, &n)){
        printf(is_bicolored(adj_list, n) ? "BICOLORABLE.\n" : "NOT BICOLORABLE.\n");
    }
    return 0;
}

File: uvaonlinejudge/10004 - Bicoloring.py
def read_test_case():
    n = int(input())
    if n == 0:
        return None
    l = int(input())
    adj_list = [[] for i in range(n)]
    for i in range(l):
        u, v = [int(j) for j in input().split()]
        adj_list[u].append(v)
        adj_list[v].append(u)
    return adj_list, n

def is_bicolored(adj_list, n):
    colors, stack = [-1] * n, [0]
    colors[0] = 1
    while len(stack) > 0:
        v = stack.pop()
        for u in adj_list[v]:
            if colors[u] == -1:
                colors[u] = 1 - colors[v]
                stack.append(u)
            elif colors[u] == colors[v]:
                return False
    return True

if __name__ == '__main__':
    for adj_list, n in iter(read_test_case, None):
        print('BICOLORABLE.' if is_bicolored(adj_list, n) else 'NOT BICOLORABLE.')
